mockfirestore/transaction.py

```diff
@@ -0,0 +1,119 @@
+from functools import partial
+import random
+from typing import Iterable, Callable
+from string import ascii_letters, digits
+from .main import MockFirestore, DocumentReference, DocumentSnapshot, Query, Timestamp
+
+MAX_ATTEMPTS = 5
+ALPHABET = list(ascii_letters) + list(digits)
+_MISSING_ID_TEMPLATE = "The transaction has no transaction ID, so it cannot be {}."
+_CANT_BEGIN = "The transaction has already begun. Current transaction ID: {!r}."
+_CANT_ROLLBACK = _MISSING_ID_TEMPLATE.format("rolled back")
+_CANT_COMMIT = _MISSING_ID_TEMPLATE.format("committed")
+
+
+class MockWriteResult:
+    def __init__(self):
+        self.update_time = Timestamp.from_now()
+
+
+class Transaction:
+    """
+    This mostly follows the model from
+    https://googleapis.dev/python/firestore/latest/transaction.html
+    """
+    def __init__(self, client: MockFirestore,
+                 max_attempts=MAX_ATTEMPTS, read_only=False):
+        self._client = client
+        self._max_attempts = max_attempts
+        self._read_only = read_only
+        self._id = None
+        self._write_ops = []
+        self.write_results = None
+
+    @property
+    def in_progress(self):
+        return self._id is not None
+
+    @property
+    def id(self):
+        return self._id
+
+    def _begin(self, retry_id=None):
+        # generate a random ID to set the transaction as in_progress
+        self._id = ''.join(random.choices(ALPHABET, k=10))
```

Comment:
Once you've rebased there's a `generate_random_string()` function you can use here.

Reply:
Thanks, will do!